### PR TITLE
feat: extract deposit requests in block observer

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes: https://github.com/stacks-network/sbtc/issues/203.

## Changes

* Rename structs and to conform more closely to our specs.
* Add a new storage trait function, `write_deposit_requests`, that stores deposit requests.
* Implement the `write_deposit_requests` function for postgres and the in-memory thing.

## Testing Information

There are unit tests for the new block observer logic and an integration test for the query.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/tests/integration/postgres.rs [hunk 1]
```diff
@@ -497,3 +498,45 @@ async fn should_return_the_same_pending_accepted_withdraw_requests_as_in_memory_
...
+        .unwrap();
+    let count =
+        sqlx::query_scalar::<_, i64>(r#"SELECT COUNT(*) FROM sbtc_signer.deposit_requests"#)
+            .fetch_one(&pool)
+            .await
+            .unwrap();
+
+    // No new records written right?
+    assert_eq!(num_rows, count as usize);
+}
```
Comment (AshtonStephens) [inline signer/tests/integration/postgres.rs:542]: Great comments on this test.

File: signer/src/storage/postgres.rs [hunk 2]
```diff
@@ -687,6 +687,94 @@ impl super::DbWrite for PgStore {
...
+            max_fee.push(req.max_fee);
+            // We need to join the addresses like this (and later split
+            // them), because handling of multi-dimensional arrays in
+            // postgres is tough. The naive approach of doing
+            // UNNEST($1::VARCHAR[][]) doesn't work, since that completely
+            // flattens the array.
+            sender_addresses.push(req.sender_addresses.join(","));
+        }
+
+        sqlx::query(
```
Comment (AshtonStephens) [inline signer/src/storage/postgres.rs:723]: Not for right now, but do we have any documentation on the structure of the SQL database? I find these very hard to follow and am really just trusting the unit tests here.

File: signer/src/block_observer.rs [hunk 3]
```diff
@@ -440,7 +642,7 @@ mod tests {
     }
 
     impl EmilyInteract for () {
-        async fn get_deposits(&mut self) -> Vec<DepositRequest> {
+        async fn get_deposits(&mut self) -> Vec<CreateDepositRequest> {
```
Comment (AshtonStephens) [inline signer/src/block_observer.rs:645]: This should likely be swapped for some autogenerated struct one day. Just flagging, for now it's fine and we may ultimately need to ship with this depending on timing.

File: sbtc/src/rpc.rs [hunk 4]
```diff
@@ -117,12 +116,13 @@ impl BtcClient {
     ///
     /// Modified from the bitcoin-core docs[1]:
     ///
-    /// This returns a "conservative" estimate, which is potentially
-    /// returns a higher fee rate and is more likely to be sufficient for
-    /// the desired target, but is not as responsive to short term drops in
-    /// the prevailing fee market. Also, the docs mention the response is
-    /// in BTC/kB, but from the comments in  bitcoin-core[2] this is really
-    /// BTC/kvB (kvB is kilo-vbyte).
+    /// Bitcoin-core has too different modes for fee rate estimation,
```
Comment (AshtonStephens) [inline sbtc/src/rpc.rs:119]:
```suggestion
    /// Bitcoin-core has two different modes for fee rate estimation,
```

File: sbtc/src/rpc.rs [hunk 5]
```diff
@@ -117,12 +116,13 @@ impl BtcClient {
...
+    /// "conservative" and "economical". We use the "conservative" estimate
+    /// because it is more likely to be sufficient for the desired target,
+    /// but is not as responsive to short term drops in the prevailing fee
+    /// market when compared to the "economical" fee rate. Also, the docs
+    /// mention the response is in BTC/kB, but from the comments in
+    /// bitcoin-core[2] this is really BTC/kvB (kvB is kilo-vbyte).
```
Comment (AshtonStephens) [inline sbtc/src/rpc.rs:125]: Love the footnotes in the docstring.

File: sbtc/src/error.rs [hunk 6]
```diff
@@ -10,6 +10,9 @@ pub enum Error {
     /// Error when creating an RPC client to bitcoin-core
     #[error("could not create RPC client to {1}; {0}")]
     BitcoinCoreRpcClient(#[source] bitcoincore_rpc::Error, String),
+    /// Error when using a BitcoinClient trait function
```
Comment (AshtonStephens) [inline sbtc/src/error.rs:13]: Don't take any action on this, but just for my own curiosity why have these in the `sbtc` crate instead of the `signer` crate?

**Review by AshtonStephens — APPROVED** [review]
Looks good to me. The structure and the naming are sound.

I find SQL hard to assess, especially when I'm not familiar with the initial database structure. Don't consider my reviews to be correct on the SQL analysis front, but I'm very in favor of committing what we have with good unit tests and finding any hiccups in this later.

One day it may make sense to write down the way we structured this database if we haven't already.

File: signer/src/storage/postgres.rs (hunk 2 above)
Comment (djordon) [reply signer/src/storage/postgres.rs:723]: Yeah we do. The migration query that created this table is here: https://github.com/stacks-network/sbtc/blob/dcbdb12183797cfe69067a789120992ab0cb4621/signer/migrations/0003__create_tables.sql#L27-L38

The rest are in that file.

File: signer/src/block_observer.rs (hunk 3 above)
Comment (djordon) [reply signer/src/block_observer.rs:645]: Yeah, I figured we'll need to switch types eventually.

File: sbtc/src/error.rs (hunk 6 above)
Comment (djordon) [reply sbtc/src/error.rs:13]: Good question. I figured Emily would want to use this functionality as well when validating deposit requests so I put the bitcoin client trait here. There is probably a better way to organize this all though.

**Review by xoloki — APPROVED** [review]
Lgtm. Small question about phantom data, but it’s not blocking anything.

File: sbtc/src/deposits.rs [hunk 7]
```diff
@@ -93,19 +97,51 @@ pub struct ParsedDepositRequest {
...
+    /// the case of a reorg, it's not clear what happens if this was
+    /// confirmed.
+    pub block_hash: Option<BlockHash>,
+    /// The number of confirmations deep from that chain tip of the bitcoin
+    /// block that includes this transaction. If `None` then this is in the
+    /// mempool. TODO(384): In the case of a reorg, it's not clear what
+    /// happens if this was confirmed.
+    pub confirmations: Option<u32>,
+    /// This is to make sure that this struct cannot be created without the
+    /// above invariants being upheld.
```
[the remark this answers is not in the document]
Comment (djordon) [reply sbtc/src/deposits.rs:119]: Oh you're right it doesn't uphold them, it just makes sure that the struct cannot be created (outside of this file) without the invariants being upheld, since you cannot set a private field of a struct outside of the file that creates it.

File: sbtc/src/rpc.rs (hunk 4 above)
Comment (djordon) [reply sbtc/src/rpc.rs:119]: Oh oops, good find. Will update.

File: sbtc/src/deposits.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (xoloki) [reply sbtc/src/deposits.rs:119]: Ah that makes waaay more sense 😂

File: sbtc/src/rpc.rs (hunk 5 above)
Comment (djordon) [reply sbtc/src/rpc.rs:125]: Done in https://github.com/stacks-network/sbtc/pull/385/commits/a2d8c5dec94c161d1b2448d4f82e407d272be2e5

Comment (djordon) [reply sbtc/src/rpc.rs:125]: Ooops, meant to write that in the above one.

File: sbtc/src/rpc.rs (hunk 4 above)
Comment (djordon) [reply sbtc/src/rpc.rs]: Done in https://github.com/stacks-network/sbtc/pull/385/commits/a2d8c5dec94c161d1b2448d4f82e407d272be2e5.
